utils/validators/formRules/index.js

```diff
@@ -0,0 +1,383 @@
+import { RBAC } from '@/config/types';
+import { HCI } from '@/config/labels-annotations';
+import isEmpty from 'lodash/isEmpty';
+import has from 'lodash/has';
+// import uniq from 'lodash/uniq';
+import cronstrue from 'cronstrue';
+
+const httpsKeys = [
+  'server-url'
+];
+
+const runValidators = (val, validators) => {
+  for (const validator of validators) {
+    const message = validator(val);
+
+    if (message) {
+      return message;
+    }
+  }
+};
+
+// "t" is the function name we use for getting a translated string
+export default function(t, opt = {}) {
+  const { displayKey = 'Value' } = opt;
+
+  // utility validators these validators only get used by other validators
+  const startDot = label => val => val?.slice(0, 1) === '.' ? t(`validation.dns.${ label }.startDot`, { key: displayKey }) : undefined;
+
+  // this one should technically be used for restricted hostnames but it doesn't look like the existing code will ever call validateHostName with restricted set to true.
+  // const endDot = label => val => val?.slice(-1) === '.' ? t(`validation.dns.${ label }.endDot`, { key: displayKey }) : undefined;
+
+  const startNumber = label => val => val?.slice(0, 1)?.match(/[0-9]/) ? t(`validation.dns.${ label }.startNumber`, { key: displayKey }) : undefined;
+
+  const startHyphen = label => val => val?.slice(0, 1) === '-' ? t(`validation.dns.${ label }.startHyphen`, { key: displayKey }) : undefined;
+
+  const endHyphen = label => val => val?.slice(-1) === '-' ? t(`validation.dns.${ label }.startHyphen`, { key: displayKey }) : undefined;
+
+  const dnsChars = (val) => {
+    const matchedChars = val?.match(/[^${ 'A-Za-z0-9-' }]/g);
+
+    if (matchedChars) {
+      return t('validation.chars', {
+        key: displayKey, count: matchedChars.length, chars: matchedChars.join(' ')
+      });
+    }
+
+    return undefined;
+  };
+
+  // the weird edge case here deals with internationalized domain names which are prepended with 'xn--'
+  // https://datatracker.ietf.org/doc/html/rfc5891#section-4.2.3.1
+  const dnsDoubleDash = val => (val?.substr(2, 2) === '--' && val?.substr(0, 2) !== 'xn') ? t(`validation.dns.doubleHyphen`, { key: displayKey }) : undefined;
+
+  const dnsIanaServiceNameDoubleDash = val => (val?.substr(2, 2) === '--' && val?.substr(0, 2) !== 'xn') ? t(`validation.dns.doubleHyphen`, { key: displayKey }) : undefined;
```

Comment:
Since there is a large number of rules, would it be cleaner if some rules were grouped into categories? like if they were defined in an object
```
{
  dns: {
    doubleDash: ...rule here,
    ianaServiceNameDoubleDash: ...rule here
    ... more dns rules
  },
  workload: {
    cronSchedule: ...cron rule
    ...more workload specific rule
  }
}
```

Reply:
Unfortunately since some of these rules are inherited from the model code, I'm stuck with the naming convention on a lot of these which precludes me from using dot notation.

There's definitely a cleaner way of doing this though, I had broken them up into separate files but had to jump through some hoops to avoid circular dependencies which ended up being more complicated than I liked. As we add additional rules, I'll definitely revisit this to find a cleaner way of organizing the rules.